### PR TITLE
V1 to v2 migration

### PR DESCRIPTION
**Pull request by MuskanThakur2002**

### CODE REVIEW

File: lib/services/coursesV2.js [hunk 1]
```diff
@@ -229,4 +229,58 @@ module.exports = class CoursesServiceV2 extends Schmervice.Service {
...
   }
+
+  async createCourses(v1Courses) {
+    const { CoursesV2, Exercises } = this.server.models();
+    const { exercisesServiceV2 } = this.server.services();
+    // eslint-disable-next-line
+    for (const course of v1Courses) {
+      const courseSingle = _.pick(course, ['id', 'name', 'logo', 'short_description']);
+      courseSingle.lang_available = 'en';
+      if (courseSingle.lang_available) {
```
Comment (jschanker) [inline lib/services/coursesV2.js:240]: If you set `courseSingle.lang_available` to `'en'` first, then the condition in the `if` statement will always be a truthy value so `courseSingle.lang_available` will always be set to `val('en').asArray().castTo('text[]')` and having the `if` won't do anything.

File: lib/services/coursesV2.js [hunk 2]
```diff
@@ -229,4 +229,58 @@ module.exports = class CoursesServiceV2 extends Schmervice.Service {
...
+
+      const courseFolderName = `${courseSingle.name}_${courseSingle.id}`;
+      if (!fs.existsSync(`curriculum_new/${courseFolderName}`)) {
+        fs.mkdirSync(`curriculum_new/${courseFolderName}`);
+      }
+      // eslint-disable-next-line
+      const courseExercises = await Exercises.query()
+        .where('course_id', courseSingle.id)
+        .orderBy('sequence_num');
+      if (courseExercises !== []) {
```
Comment (jschanker) [inline lib/services/coursesV2.js:254]: `courseExercises !== []` will always evaluate to `true` in JavaScript because you're comparing reference types (like comparing memory addresses rather than contents of list as would happen when you use Python's `!=`).  In particular, in JavaScript, `[] !== []` evaluates to `true`!  For **Python**, this would be like using `[] is not []` (https://docs.python.org/3/reference/expressions.html#is-not) which would be `True` instead of `[] != []`, which evaluates to `False`.  Instead, you could use `courseExercises.length !== 0` here to check that the Array is nonempty.

But actually, is this `if` statement even necessary?  If `courseExercises` is an empty Array, then the body of the `for` loop won't be run anyway.  So you could get rid of the `if` here.

File: lib/services/coursesV2.js [hunk 3]
```diff
@@ -229,4 +229,58 @@ module.exports = class CoursesServiceV2 extends Schmervice.Service {
...
+          } else {
+            let newName = singleExercise.slug.split('__');
+            newName = newName[newName.length - 1].split('/').join('_').split('_');
+            newName = newName.slice(0, newName.length - 1).join('_');
+            exerciseData.name = newName;
+          }
+          let filePath = singleExercise.slug.split('__').join('/');
+          const fileRealPath = filePath.split('_').pop();
+          filePath = filePath.split('_');
+          filePath = `${filePath.slice(0, filePath.length - 1).join('_')}.${fileRealPath}`;
```
Comment (jschanker) [inline lib/services/coursesV2.js:269]: I think that all this splitting, joining, slicing, and popping obscures how you're getting the `newName` and `filePath` from `singleExercise.slug`.  If you don't mind using regular expressions, how about something like this instead:
```suggestion
          // slug example: course_json__lesson1_json 
          // filePath: course_json/lesson1.json
          const filePath = singleExercise.slug.replace(/__/g, '/') // all __ replaced with /
              .replace(/_([^_]*)$/, '.$1'); // last _ replaced with a .
          exerciseData.name = singleExercise.parent_exercise_id == null ?
              singleExercise.name :
               // name = lesson1, between last '/' and '.'
              filePath.substring(filePath.lastIndexOf('/')+1, filePath.lastIndexOf('.'));    
```
This doesn't exactly coincide with what you have in some cases (e.g., if there were no underscores), but I think it probably coincides for the slugs that are valid.

In case you're not too familiar regular expressions, the `g` flag in the first regular expression means global so all occurrences of `__` will be replaced with `/`.  The `[^_]` matches a single **non**-underscore ( _ ), `*` means zero or more, the `$` indicates the end of the input and the parentheses around `[^_]*` are capturing parentheses that save the match for use in the replacement.  The match in the first capturing parentheses, can then be recalled using `$1`.  So `_([^_]*)$` matches (the last) underscore followed by all of the non-underscore characters up until the end of the string and replaces it with a `.` followed by the non-underscore characters to the end.  See also https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Groups_and_Ranges and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Cheatsheet#boundary-type_assertions.

File: lib/services/coursesV2.js [hunk 4]
```diff
@@ -229,4 +229,58 @@ module.exports = class CoursesServiceV2 extends Schmervice.Service {
...
+        for (const singleExercise of courseExercises) {
+          const exerciseData = {};
+          if (singleExercise.parent_exercise_id == null) {
+            exerciseData.name = singleExercise.name;
+          } else {
+            let newName = singleExercise.slug.split('__');
+            newName = newName[newName.length - 1].split('/').join('_').split('_');
+            newName = newName.slice(0, newName.length - 1).join('_');
+            exerciseData.name = newName;
+          }
```
Comment (jschanker) [inline lib/services/coursesV2.js:265]: I think that all this splitting, joining, and slicing obscures how you're getting the `newName` and `filePath` from `singleExercise.slug`.  How about something like this instead?
```suggestion
          // slug_example: course_json__lessons__lesson1/part1_json
          exerciseData.name = singleExercise.parent_exercise_id == null ?
              singleExercise.name :
              // get between last '__' and '_' (lesson1/part1)
              singleExercise.slug.substring(singleExercise.slug.lastIndexOf('__') + 2,
              singleExercise.slug.lastIndexOf('_'))
              .replace('/\//g', '_'); // replace all / with _ (lesson1_part1)  
```
This assumes that there is at least one `__` and that if there is a `/` in `singleExercise.slug`, then the last `_` comes after the last `/`.

File: lib/routes/courses.js [hunk 5]
```diff
@@ -158,8 +158,6 @@ module.exports = [
             );
             course[0].exercises[exercise].sequence_num = sequence;
             course[0].exercises[exercise].content_type = 'exercise';
-            const exerciseContent = JSON.parse(JSON.parse(course[0].exercises[exercise].content));
```
Comment (pysaquib) [inline lib/routes/courses.js:161]: @MuskanThakur2002 Why was this removed?

File: lib/routes/coursesManagement.js [hunk 6]
```diff
@@ -413,4 +413,28 @@ module.exports = [
       },
     },
   },
+  {
+    method: 'POST',
+    path: '/courses/v1_to_v2_migration',
```
Comment (pysaquib) [inline lib/routes/coursesManagement.js:418]: @MuskanThakur2002 Did you remove the part where course id was copied into v2 from v1?

File: lib/services/courses.js [hunk 7]
```diff
@@ -196,11 +196,15 @@ module.exports = class CoursesService extends Schmervice.Service {
     }
   }
 
-  async getAllCourses(txn = null) {
+  async getAllCourses(criteria = null) {
```
Comment (pysaquib) [inline lib/services/courses.js:199]: `criteria` can be a confusing variable. You can use `courseType` instead.

File: lib/services/coursesV2.js [hunk 8]
```diff
@@ -43,6 +43,22 @@ module.exports = class CoursesServiceV2 extends Schmervice.Service {
         .modify((builder) => {
           builder.orderBy('exercisesV2.sequence_num');
         });
+      _.map(courseExercises[0].exercisesV2, (exercise) => {
+        const courseName = `${courseExercises[0].name}_${courseExercises[0].id}_${exercise.name}_${lang}.json`;
+        const modifiedFile = `curriculum_new/${courseExercises[0].name}_${courseExercises[0].id}/PARSED_CONTENT/MODIFIED_FILES/${courseName}`;
```
Comment (pysaquib) [inline lib/services/coursesV2.js:48]: You can do this `const baseDir = curriculum_new/${courseExercises[0].name}_${courseExercises[0].id}/PARSED_CONTENT` and

`const modifiedFile = ${baseDir}/MODIFIED_FILES/${courseName}`
`const propertiesFile = ${baseDir}/PROPERTIES_FILES/${courseName}`

File: lib/services/coursesV2.js [hunk 9]
```diff
@@ -229,4 +245,51 @@ module.exports = class CoursesServiceV2 extends Schmervice.Service {
       return [errorHandler(err), null];
     }
   }
+
+  async createCourses(v1Courses) {
```
Comment (pysaquib) [inline lib/services/coursesV2.js:249]: What is the param `v1Courses` refer to here?

**Review by pysaquib — COMMENTED** [review]
Check comments

File: lib/services/courses.js (hunk 7 above)
Comment (MuskanThakur2002) [reply lib/services/courses.js:199]: okay

File: lib/services/coursesV2.js (hunk 8 above)
Comment (MuskanThakur2002) [reply lib/services/coursesV2.js:48]: okay

File: lib/services/coursesV2.js (hunk 9 above)
Comment (MuskanThakur2002) [reply lib/services/coursesV2.js:249]: courseV1 details

File: lib/routes/coursesManagement.js (hunk 6 above)
Comment (MuskanThakur2002) [reply lib/routes/coursesManagement.js:418]: yes
